auditing/timescaledb.go

```diff
@@ -0,0 +1,345 @@
+		migrator.Migrations(
+			initialSchema,
+		),
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := m.Migrate(a.db.DB); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *timescaleAuditing) Flush() error {
+	return nil
+}
+
+func (a *timescaleAuditing) Index(entry Entry) error {
+	if entry.Timestamp.IsZero() {
+		return errors.New("timestamp is not set")
+	}
+
+	q := "INSERT INTO traces (timestamp, entry) VALUES (:timestamp, :entry)"
+
+	e, err := json.Marshal(entry)
+	if err != nil {
+		return fmt.Errorf("error marshaling entry: %w", err)
+	}
+
+	row := timescaledbRow{
+		Timestamp: entry.Timestamp.UTC(),
+		Entry:     e,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	_, err = a.db.NamedExecContext(ctx, q, row)
+	if err != nil {
+		return fmt.Errorf("unable to index audit trace: %w", err)
+	}
+
+	return nil
+}
+
+func (a *timescaleAuditing) Search(ctx context.Context, filter EntryFilter) ([]Entry, error) {
+	var (
+		where     []string
+		values    = map[string]interface{}{}
+		addFilter = func(field string, value any, op sqlCompOp) error {
+			if reflect.ValueOf(value).IsZero() {
+				return nil
+			}
+
+			values[field] = value
+
+			switch op {
+			case equals:
+				where = append(where, fmt.Sprintf("entry ->> '%s'=:%s", field, field))
+			case like:
+				where = append(where, fmt.Sprintf("entry ->> '%s' like '%%' || :%s || '%%'", field, field))
+			case phrase:
+				// the additional "like" match allows matching partial words, too
+				where = append(where, fmt.Sprintf("ts @@ websearch_to_tsquery('simple', '$$' || :%s || '$$') or entry ->> '%s' like '%%' || :%s || '%%'", field, field, field))
+			default:
+				return fmt.Errorf("comp op not known")
+			}
+
+			return nil
+		}
+	)
+
+	if err := addFilter("body", filter.Body, phrase); err != nil {
+		return nil, err
+	}
+	if err := addFilter("component", filter.Component, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("detail", filter.Detail, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("error", filter.Error, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("forwardedfor", filter.ForwardedFor, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("path", filter.Path, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("phase", filter.Phase, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("remoteaddr", filter.RemoteAddr, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("rqid", filter.RequestId, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("statuscode", filter.StatusCode, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("tenant", filter.Tenant, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("type", filter.Type, equals); err != nil {
+		return nil, err
+	}
+	if err := addFilter("user", filter.User, equals); err != nil {
+		return nil, err
+	}
+
```

Comment:
this must all match to the Entry json annotations, easy to make mistakes here or miss some of them. There might be a smarter way, but unsure.

Reply:
Only other approach that comes to my mind is to put the field names into constants and build a generic `map[string]any` with them, which will be JSON-serialized instead of using the struct with JSON annotations. Not sure if it's better.